libupload.groovy

```diff
@@ -1,4 +1,92 @@
+
+    def meta = readJSON(text: shwrapCapture("cosa meta --arch=${basearch} --dump"))
+    def replicators = [:]
+    def credentials
+
+    credentials = [file(variable: "ALIYUN_IMAGE_UPLOAD_CONFIG",
+                        credentialsId: "aliyun-image-upload-config")]
+    if (meta.aliyun) {
+        def creds = credentials
+        replicators["☁️ ⬆️ :aliyun"] = {
+            withCredentials(creds) {
+                utils.syncCredentialsIfInRemoteSession(["ALIYUN_IMAGE_UPLOAD_CONFIG"])
+                def c = pipecfg.clouds.aliyun
+                def extraArgs = []
+                if (c.regions) {
+                    extraArgs = "--regions" +  c.regions.join(" ")
+                }
+                if (c.public) {
+                    extraArgs += "--public"
+                }
+                shwrap("""
+                coreos-assembler aliyun-replicate \
+                    --build=${buildID} \
+                    --config=\${ALIYUN_IMAGE_UPLOAD_CONFIG} \
+                    ${extraArgs.join(' ')}
+                """)
+            }
+        }
+    }
+    if (meta.amis) {
+        replicators["☁️ ⬆️ :aws"] = {
+            def replicated = false
+            def extraArgs = []
+            def ids = ['aws-build-upload-config', 'aws-govcloud-image-upload-config']
+            def c = pipecfg.clouds.aws
+            if (c.regions) {
+                extraArgs = "--regions" +  c.regions.join(" ")
+            }
+            for (id in ids) {
+                tryWithCredentials([file(variable: 'AWS_CONFIG_FILE', credentialsId: id)]) {
+                    shwrap("""
+                    cosa aws-replicate \
+                        --build=${buildID} \
+                        --arch=${basearch} \
+                        --source-region ${c.primary_region} \
+                        --credentials-file=\${AWS_CONFIG_FILE} \
+                        --log-level=INFO \
+                        ${extraArgs.join(' ')}
+                    """)
+                    replicated = true
+                }
+            }
+            if (!replicated) {
+                        error("AWS Replication asked for but no credentials exist")
+            }
+        }
+    }
+    credentials = [file(variable: "POWERVS_IMAGE_UPLOAD_CONFIG",
+                        credentialsId: "powervs-image-upload-config")]
+    if (meta.powervs) {
+        def creds = credentials
+        replicators["☁️ ⬆️ :powervs"] = {
+            withCredentials(creds) {
+                utils.syncCredentialsIfInRemoteSession(["POWERVS_IMAGE_UPLOAD_CONFIG"])
+                def c = pipecfg.clouds.powervs
+                // for powervs in RHCOS the images are uploaded to a bucket in each
+                // region that is uniquely named with the region as a suffix
+                // i.e. `rhcos-powervs-images-us-east`
+                def regions = []
+                if (c.regions) {
+                     regions = c.regions.join(" ")
+                }
```

Comment:
I wonder if it's possible to autodetect here like we do for the other clouds?

Reply:
I'm still waiting for a test secret to see how things go. =/. I will try as son as I got it.

Reply:
Let's not wait on the credentials for this. Let's leave this code in for now but also document it in `docs/config.yaml`. We can post any followup fixes/improvements that we find in the coming days.

I'd like to get this PR merged sooner than later so we can start really testing it out (in the FCOS pipeline).